File: swift/windowing-system/Sources/WindowingSystem/WindowingSystem.swift
struct Position {
    var x = 0
    var y = 0

    mutating func moveTo(newX: Int, newY: Int) {
        x = newX
        y = newY
    }
}

struct Size {

    var width = 80
    var height = 60

    mutating func resize(newWidth: Int, newHeight: Int) {
        width = newWidth
        height = newHeight
    }

}

class Window {

    var title = "New Window"
    var screenSize = Size(width: 800, height: 600)
    var size = Size()
    var position = Position()
    var contents: String?

    init() {
    }

    init(_ t: String, _ sz: Size, _ pos: Position, _ cont: String?) {
        title = t
        size = sz
        position = pos
        contents = cont
    }


    func resize(to s: Size) {
        size.width = min(max(1, s.width), screenSize.width - position.x)
        size.height = min(max(1, s.height), screenSize.height - position.y)
    }

    func move(to pos: Position) {
        position.x = min(max(0, pos.x), screenSize.width - size.width)
        position.y = min(max(0, pos.y), screenSize.height - size.height)
    }

    func update(title t: String) {
        title = t
    }

    func update(text: String?) {
        contents = text
    }

    func display() -> String {
        "\(title)\nPosition: (\(position.x), \(position.y)), Size: (\(size.width) x \(size.height))\n\(contents ?? "[This window intentionally left blank]")\n"
    }

}

let mainWindow = Window("Main Window", Size(width: 400, height: 300), Position(x: 100, y: 100), "This is the main window")

File: javascript/windowing-system/windowing-system.js
export function Size(width = 80, height = 60) {
  this.width = width;
  this.height = height;
}

Size.prototype.resize = function (newWidth, newHeight) {
  this.width = newWidth;
  this.height = newHeight;
}

export function Position(x = 0, y = 0) {
  this.x = x;
  this.y = y;
}

Position.prototype.move = function (newX, newY) {
  this.x = newX;
  this.y = newY;
}

export class ProgramWindow {
  constructor() {
    this.screenSize = new Size(800, 600);
    this.size = new Size();
    this.position = new Position();
  }

  resize(newSize) {
    if (newSize.width < 1) newSize.width = 1;
    if (newSize.height < 1) newSize.height = 1;
    if (newSize.width + this.position.x > this.screenSize.width) newSize.width = this.screenSize.width - this.position.x;
    if (newSize.height + this.position.y > this.screenSize.height) newSize.height = this.screenSize.height - this.position.y;
    this.size.height = newSize.height;
    this.size.width = newSize.width;
  }

  move(newPosition) {
    if (newPosition.x < 0) newPosition.x = 0;
    if (newPosition.y < 0) newPosition.y = 0;
    if (newPosition.x + this.size.width > this.screenSize.width) newPosition.x = this.screenSize.width - this.size.width;
    if (newPosition.y + this.size.height > this.screenSize.height) newPosition.y = this.screenSize.height - this.size.height;
    this.position.x = newPosition.x;
    this.position.y = newPosition.y;
  }
}

export function changeWindow(window) {
  window.size = new Size(400, 300);
  window.position = new Position(100, 150);
  return window;
}
